### DISCUSSION TITLE
mail check and notifications

### DISCUSSION THREAD

**Discussion by roccoblues** [neomutt/neomutt#4053] [Development]
**Issues:**

* Recent refactoring broke the mail check for the current mailbox after running an external command (https://github.com/neomutt/neomutt/issues/4048#issuecomment-1734103362).

* [dlg_index()](https://github.com/neomutt/neomutt/blob/main/index/dlg_index.c#L1216-L1230) and [dlg_pager()](https://github.com/neomutt/neomutt/blob/main/pager/dlg_pager.c#L448-L462)) duplicate the code to check for new mail and trigger the corresponding actions.

* [mutt_mailbox_check()](https://github.com/neomutt/neomutt/blob/main/mutt_mailbox.c#L165) claims to check all mailboxes for new mail but it skips the [current one](https://github.com/neomutt/neomutt/blob/main/mutt_mailbox.c#L118). I guess that is because the current mailbox is checked in the event loops but it's still confusing.

**Naive refactoring idea:**

Remove the mail check from the event loops and add a `$timeout` handler that according to `$mail_check` and `$mail_check_stats_interval` triggers the checks. That also sends out a `NT_MAILBOX` notification if we have new mail.
That can be picked up by a handler that runs all the actions: beep, show the "new mail" message and run `new_mail_command`.

**Comment by flatcap** [comment]
Yeah, I think this layered approach is the way forward.
I'll write up my notes tomorrow.
Meanwhile, a diagram:

<img src="https://raw.githubusercontent.com/neomutt/gfx/main/seq/new-mail3.svg">

[source](https://github.com/neomutt/gfx/tree/main/seq) [generator](https://sequencediagram.org/)

**Comment by gahr** [comment]
> Remove the mail check from the event loops

+1

**Comment by roccoblues** [comment]
From an IRC discussion:

**When do we need to check all mailboxes?**

1. on demand `<check-mailboxes>`
2. on timer, `$mail_check`
3. on startup?

**When do we need to check a specific mailbox?**

1. on switching mailboxes:
  a. the old mailbox to sync changes
  b. the new mailbox to check for new mail

The current mail checks after running external commands (`<shell-escape>`) can be removed when `<check-mailboxes>` checks _all_ mailboxes: `<shell-escape>mbsync -a && notmuch new<enter><check-mailboxes>`

**Comment by roccoblues** [comment]
Playing around with a new-mail observer: https://github.com/neomutt/neomutt/pull/4077
